data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockHeader.java

```diff
@@ -74,17 +108,64 @@ public GetBlockHeader(
       })
   @Override
   public void handle(@NotNull final Context ctx) throws Exception {
-    final Map<String, String> pathParams = ctx.pathParamMap();
-    final SafeFuture<Optional<ObjectAndMetaData<BlockHeader>>> future =
-        chainDataProvider.getBlockHeader(pathParams.get(PARAM_BLOCK_ID));
-    handleOptionalResult(ctx, future, this::handleResult, SC_NOT_FOUND);
+    adapt(ctx);
+  }
+
+  @Override
+  public void handleRequest(RestApiRequest request) throws JsonProcessingException {
+    final SafeFuture<Optional<BlockAndMetaData>> future =
+        chainDataProvider.getBlockAndMetaData(request.getPathParameter(PARAMETER_BLOCK_ID));
+
+    request.respondAsync(
+        future.thenApply(
+            maybeBlockAndMetaData -> {
+              if (maybeBlockAndMetaData.isEmpty()) {
+                return AsyncApiResponse.respondNotFound();
+              }
+
+              final BlockAndMetaData blockAndMetaData = maybeBlockAndMetaData.get();
+              return AsyncApiResponse.respondOk(blockAndMetaData);
+            }));
```

Comment:
this is starting to look really common, i wonder if we need an abstract base class for some of these?